hathor/utils/api.py

```diff
@@ -0,0 +1,54 @@
+#  Copyright 2023 Hathor Labs
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from __future__ import annotations
+
+from typing import List, Optional, TypeVar
+
+from pydantic import Field, ValidationError, validator
+from twisted.web.http import Request
+
+from hathor.api_util import get_args
+from hathor.utils.pydantic import BaseModel
+
+T = TypeVar('T')
+
+
+def _list_to_single_item(v: List[T]) -> Optional[T]:
+    assert len(v) <= 1, 'expected one value at most'
+
+    return None if not len(v) else v[0]
+
+
+class QueryParams(BaseModel):
+    _list_to_single_item_validator = validator('*', pre=True, allow_reuse=True)(_list_to_single_item)
+
+    @classmethod
+    def from_request(cls, request: Request) -> QueryParams | ErrorResponse:
+        raw_args = get_args(request).items()
+        args = {k.decode('utf8'): v for k, v in raw_args}
```

Comment:
As this is a generic tool, should we get encoding from the headers?

Reply:
Actually I'm not sure how this works in Twisted. This dictionary comes from the `args` property in the `Request` class. According to the [docs](https://docs.twisted.org/en/stable/api/twisted.web.http.Request.html#args):

> A mapping of decoded query argument names as [bytes](https://docs.python.org/3/library/stdtypes.html#bytes) to corresponding query argument values as [list](https://docs.python.org/3/library/stdtypes.html#list)s of [bytes](https://docs.python.org/3/library/stdtypes.html#bytes). For example, for a URI with foo=bar&foo=baz&quux=spam as its query part args will be {b'foo': [b'bar', b'baz'], b'quux': [b'spam']}.

It mentions "decoded query argument names as bytes". So how these bytes are encoded is not clear to me. Are you sure this is the same as the encoding from the headers? From the `Content-Type`?

Reply:
I feel that Twisted just pass along the bytes received by the request, i.e., it does nothing to the data. If that's correct, you can try use the content-type and use utf-8 as a fallback.

Reply:
Done!